java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java

```diff
@@ -0,0 +1,144 @@
+package co.worklytics.psoxy.gateway.impl.oauth;
+
+import co.worklytics.psoxy.gateway.ConfigService;
+import co.worklytics.psoxy.gateway.RequiresConfiguration;
+import co.worklytics.psoxy.gateway.SourceAuthStrategy;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.api.client.http.*;
+import com.google.auth.Credentials;
+import com.google.auth.oauth2.AccessToken;
+import com.google.auth.oauth2.OAuth2CredentialsWithRefresh;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.extern.java.Log;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+
+/**
+ * source auth strategy to authenticate using a short-lived OAuth 2.0 access token which must be
+ * periodically refreshed.
+ *   Options for refresh method are configured by
+ *
+ * A new access token will be retrieved for every psoxy instance that spins up; as well as when the
+ * current one expires.  We'll endeavor to minimize the number of token requests by sharing this
+ * states across API requests
+ *
+ * If the source API you're connecting to offers long-lived access tokens (or does not offer refresh
+ * tokens), you may opt for the access-token only strategy:
+ * @see OAuthAccessTokenSourceAuthStrategy
+ *
+ */
+@Log
+@NoArgsConstructor(onConstructor_ = @Inject)
+public class OAuthRefreshTokenSourceAuthStrategy implements SourceAuthStrategy {
+
+    @Getter
+    private final String configIdentifier = "oauth2_refresh_token";
+
+
+    //q: should we put these as config properties? creates potential for inconsistent configs
+    // eg, orphaned config properties for SourceAuthStrategy not in use; missing config properties
+    //  expected by this
+    public enum ConfigProperty implements ConfigService.ConfigProperty {
+        REFRESH_ENDPOINT,
+        CLIENT_ID,
+        GRANT_TYPE,
+    }
+
+    @Inject OAuth2CredentialsWithRefresh.OAuth2RefreshHandler refreshHandler;
+
+    @Override
+    public Set<ConfigService.ConfigProperty> getRequiredConfigProperties() {
+        Stream<ConfigService.ConfigProperty> propertyStream = Arrays.stream(ConfigProperty.values());
+        if (refreshHandler instanceof RequiresConfiguration) {
+            propertyStream = Stream.concat(propertyStream,
+                ((RequiresConfiguration) refreshHandler).getRequiredConfigProperties().stream());
+        }
+        return propertyStream.collect(Collectors.toSet());
+    }
+
+
+    @Override
+    public Credentials getCredentials(Optional<String> userToImpersonate) {
+        return OAuth2CredentialsWithRefresh.newBuilder()
+            //TODO: pull an AccessToken from some cached location or something? otherwise will
```

Comment:
:+1:  Yeah, a cache with the token expiration as expiration; it could be invalidated if a new access token is requested as well